src/client.rs

```diff
@@ -0,0 +1,267 @@
+                    "Substitution key was not found.".to_string(),
+                ));
+            }
+            res = subbed_str;
+        }
+
+        // TODO May be a better way to handle this, but {} are not valid uri chars
+        if res.contains('{') && res.contains('}') {
+            return Err(TurnkeyError::OtherError(
+                "Did not substitute all keys.".to_string(),
+            ));
+        }
+        Ok(res)
+    }
+    pub fn construct_url(
+        &self,
+        uri: String,
+        query: QueryShape,
+        substitution: SubstitutionShape,
+    ) -> TurnkeyResult<String> {
+        let path = Turnkey::substitute_path(uri, substitution)?;
+        // TODO: Might need to check if path starts with "/", not sure how its called
+        let mut url = format!("{}{}", self.base_url, path);
+
+        // TODO: This can all be cleaned up/simplified with Url crate if needed
+        if !query.is_empty() {
+            url = format!("{}?", url);
+        }
+        for (key, val) in query.iter() {
+            match val {
+                QueryValueShape::Single(item) => {
+                    url = format!("{}&{}={}", url, key, item);
+                }
+                QueryValueShape::Array(items) => {
+                    for item in items {
+                        url = format!("{}&{}={}", url, key, item);
+                    }
+                }
+            }
+        }
+        Ok(url)
+    }
+
+    pub async fn request<ResponseData>(
+        &self,
+        request_input: RequestInput,
+    ) -> TurnkeyResult<ResponseData>
+    where
+        ResponseData: DeserializeOwned,
+    {
+        // TODO: unwrap_or_default correct here? or make construct_url take in an Option
+        let url = self.construct_url(
+            request_input.uri,
+            request_input.query.unwrap_or_default(),
+            request_input.substitution.unwrap_or_default(),
+        )?;
+
+        let sealed_request_input = SealedRequestInput {
+            body: request_input.body.unwrap_or_default(),
+            public_key_hex: None,
+            private_key_hex: None,
+        };
+        let sealed_request_output = self
+            .seal_and_stamp_request_body(sealed_request_input)
+            .await?;
+
+        let response = self
+            .client
+            .post(&url)
+            .headers({
+                let mut headers = reqwest::header::HeaderMap::new();
+                if let Ok(x_stamp_header) =
+                    HeaderValue::try_from(sealed_request_output.x_stamp.clone())
+                {
+                    headers.insert("X-Stamp", x_stamp_header);
+                }
+                match request_input.headers {
+                    Some(input_headers) => {
+                        for (_, value) in input_headers.iter() {
+                            if let Ok(val_header) = HeaderValue::try_from(value) {
+                                headers.insert("test keyname", val_header);
```

Comment:
do we need this in the case that there are other headers the user would like to include? or purely for testing?

Reply:
cc @joshhchun